src/platform/datapath_epoll.c

```diff
@@ -2080,13 +2080,35 @@ CxPlatSocketSendInternal(
     size_t TotalMessagesCount;
 
     CXPLAT_DBG_ASSERT(Socket != NULL && RemoteAddress != NULL && SendData != NULL);
+    CXPLAT_DBG_ASSERT(SendData->SentMessagesCount <= CXPLAT_MAX_BATCH_SEND);
 
     static_assert(CMSG_SPACE(sizeof(struct in6_pktinfo)) >= CMSG_SPACE(sizeof(struct in_pktinfo)), "sizeof(struct in6_pktinfo) >= sizeof(struct in_pktinfo) failed");
     char ControlBuffer[CMSG_SPACE(sizeof(struct in6_pktinfo)) + CMSG_SPACE(sizeof(int))] = {0};
 
     ProcNumber = CxPlatProcCurrentNumber() % Socket->Datapath->ProcCount;
     SocketContext = &Socket->SocketContexts[Socket->HasFixedRemoteAddress ? 0 : ProcNumber];
 
+    ProcContext = &Socket->Datapath->ProcContexts[ProcNumber];
+
+    uint32_t TotalSize = 0;
+    for (size_t i = SendData->SentMessagesCount; i < SendData->BufferCount; ++i) {
+        SendData->Iovs[i].iov_base = SendData->Buffers[i].Buffer;
+        SendData->Iovs[i].iov_len = SendData->Buffers[i].Length;
+        TotalSize += SendData->Buffers[i].Length;
+    }
+
+    if (!IsPendedSend) {
+        QuicTraceEvent(
+            DatapathSend,
+            "[data][%p] Send %u bytes in %hhu buffers (segment=%hu) Dst=%!ADDR!, Src=%!ADDR!",
+            Socket,
+            TotalSize,
+            SendData->BufferCount,
+            SendData->Buffers[SendData->SentMessagesCount].Length,
```

Comment:
Is it guaranteed that Count == 0 when IsPendedSend is FALSE? Worth another DBG assert?

Reply:
Yeah I can add that. The only way for a non pended send to get > 0 here is if the core code sets that somehow, but worth checking.